src/metabase/sample_data.clj

```diff
@@ -3,42 +3,90 @@
+            [metabase.plugins :as plugins]
             [metabase.sync :as sync]
+            [metabase.util.files :as u.files]
             [metabase.util.i18n :refer [trs]]
-            [toucan.db :as db]))
+            [toucan.db :as db])
+  (:import java.net.URL))
 
 (def ^:private ^String sample-database-name     "Sample Database")
 (def ^:private ^String sample-database-filename "sample-database.db.mv.db")
 
-(defn- db-details []
+;; Reuse the plugins directory for the destination to extract the sample database because it's pretty much guaranteed
+;; to exist and be writable.
+(defn- target-path
+  []
+  (u.files/append-to-path (plugins/plugins-dir) sample-database-filename))
+
+(defn- extract-sample-database!
+  []
+  (u.files/with-open-path-to-resource [sample-db-path sample-database-filename]
+    (u.files/copy-file! sample-db-path (target-path))))
+
+(defn- process-sample-db-path
+  [base-path]
+  (-> base-path
+      (str/replace #"\.mv\.db$" "")        ; strip the .mv.db suffix from the path
+      (str/replace #"%20" " ")             ; for some reason the path can get URL-encoded and replace spaces with `%20`;
+                                           ;   this breaks things so switch them back to spaces
+      (str ";USER=GUEST;PASSWORD=guest"))) ; specify the GUEST user account created for the DB
+
+(defn- jar-db-details
+  [resource]
+  (-> (.getPath ^URL resource)
+      (str/replace #"^file:" "zip:") ; to connect to an H2 DB inside a JAR just replace file: with zip: (this doesn't
+                                     ;   do anything when running from the Clojure CLI, which has no `file:` prefix
+      process-sample-db-path))
+
+(defn- extracted-db-details
+  []
+  (-> (str "file:" (plugins/plugins-dir) "/" sample-database-filename)
+      process-sample-db-path))
+
+(defn- try-to-extract-sample-database!
+  "Tries to extract the sample database out of the JAR (for performance) and then returns a db-details map
+   containing a path to the copied database."
+  []
   (let [resource (io/resource sample-database-filename)]
     (when-not resource
       (throw (Exception. (trs "Sample database DB file ''{0}'' cannot be found."
                               sample-database-filename))))
-    {:db (-> (.getPath resource)
-             (str/replace #"^file:" "zip:") ; to connect to an H2 DB inside a JAR just replace file: with zip: (this doesn't do anything when running from the Clojure CLI, which has no `file:` prefix)
-             (str/replace #"\.mv\.db$" "")  ; strip the .mv.db suffix from the path
-             (str/replace #"%20" " ") ; for some reason the path can get URL-encoded and replace spaces with `%20`; this breaks things so switch them back to spaces
-             (str ";USER=GUEST;PASSWORD=guest"))})) ; specify the GUEST user account created for the DB
+    {:db
+     (if-not (:temp (plugins/plugins-dir-info))
+       (do
+        (extract-sample-database!)
+        (extracted-db-details))
```

Comment:
Consider returning the details from `extract-sample-database!` (or computing the details from it).